crates/ruma-server-util/src/authorization.rs

```diff
@@ -0,0 +1,198 @@
+        Self::new()
+    }
+}
+
+impl Credentials for XMatrix {
+    const SCHEME: &'static str = "X-Matrix";
+
+    fn decode(value: &HeaderValue) -> Option<Self> {
+        debug_assert!(
+            value.as_bytes().starts_with(b"X-Matrix "),
+            "HeaderValue to decode should start with \"X-Matrix ..\", received = {:?}",
+            value,
+        );
+
+        let parameters =
+            std::str::from_utf8(&value.as_bytes()["X-Matrix ".len()..]).ok()?.trim_start();
+
+        let mut builder: XMatrixBuilder = Default::default();
+
+        for entry in parameters.split_terminator(',') {
+            let (name, value) = entry.split_once('=')?;
+
+            // It's not at all clear why some fields are quoted and others not in the spec,
+            // let's simply accept either form for every field.
+            let value =
+                value.strip_prefix('"').and_then(|rest| rest.strip_suffix('"')).unwrap_or(value);
+
+            builder = match name {
+                "origin" => builder.origin(value.try_into().ok()?),
+                "destination" => builder.destination(value.try_into().ok()?),
+                "key" => builder.key(value.try_into().ok()?),
+                "sig" => builder.sig(value.to_owned()),
+                name => {
+                    debug!("Unknown field {} found in X-Matrix Authorization header", name);
+                    builder
+                }
+            };
+        }
+
+        builder.build()
+    }
+
+    fn encode(&self) -> HeaderValue {
+        if let Some(destination) = &self.destination {
+            format!(
+                "X-Matrix origin=\"{}\",destination=\"{}\",key=\"{}\",sig=\"{}\"",
+                self.origin, destination, self.key, self.sig
+            )
+        } else {
+            format!("X-Matrix origin=\"{}\",key=\"{}\",sig=\"{}\"", self.origin, self.key, self.sig)
+        }
+        .try_into()
+        .expect("")
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use headers::{authorization::Credentials, HeaderValue};
+    use ruma_common::ServerName;
+
+    use crate::authorization::XMatrixBuilder;
+
+    use super::XMatrix;
+
+    #[test]
+    fn xmatrix_auth_pre_1_3() {
+        let header = HeaderValue::from_static(
+            "X-Matrix origin=\"origin.hs.example.com\",key=\"ed25519:key1\",sig=\"ABCDEF...\"",
+        );
+        let origin = "origin.hs.example.com".try_into().unwrap();
+        let key = "ed25519:key1".try_into().unwrap();
+        let sig = "ABCDEF...".to_string();
+        let credentials: XMatrix = Credentials::decode(&header).unwrap();
+        assert_eq!(credentials.origin, origin);
+        assert_eq!(credentials.destination, None);
+        assert_eq!(credentials.key, key);
+        assert_eq!(credentials.sig, sig);
+
+        let credentials = XMatrixBuilder::new().origin(origin).key(key).sig(sig).build().unwrap();
+
+        assert_eq!(credentials.encode(), header);
+    }
+
+    #[test]
+    fn xmatrix_auth_post_1_3() {
```

Comment:
```suggestion
    fn xmatrix_auth_1_3() {
```

post implies "after" 1.3, so 1.4+